mlrun/api/utils/projects/manager.py

```diff
@@ -0,0 +1,267 @@
+import collections
+import typing
+
+import humanfriendly
+import sqlalchemy.orm
+
+import mlrun.api.db.session
+import mlrun.api.schemas
+import mlrun.api.utils.periodic
+import mlrun.api.utils.projects.consumers.base
+import mlrun.api.utils.projects.consumers.nop
+import mlrun.api.utils.projects.consumers.nuclio
+import mlrun.config
+import mlrun.errors
+import mlrun.utils
+import mlrun.utils.singleton
+from mlrun.utils import logger
+
+
+class ProjectsManager(metaclass=mlrun.utils.singleton.Singleton):
```

Comment:
You're mixing two things here (explaining in my current terminology):
we have Nuclio and MLRun (and Zebo in the future) which are all **consumers** of the projects resource.
**Any of them** can be configured to be the **master consumer**
The Projects Manager is the entity responsible to perform actions by the configured "hierarchy".
So I'm ok with changing their "type" name to follower, and then there is one follower that is configured as the leader (though TBH while "master-consumer" makes sense "grammatically", "leader-follower" does not, IMO)
but Leader can't be the name of the Projects Manager, this is another thing, how you wanna call it ?

Reply:
I actually understand what you did and didn't mix things up.
I just think that there is a slightly artificial separation here and it makes it look like a master and a controller are not the same thing, but that's a false pretense. I'll explain - if mlrun is not the "master" then its "manager" shouldn't be operational. In other words - the manager is not something that is always here and it just "happens" to have mlrun configured as master. If some other component would be the master - mlrun's "manager" should not be operation, it would not be in charge of syncing, the master would. These are not 2 separate things - the "manager" is what makes the master a "master". Making it a wrapping / separate object is cool, but its part of the same mechanism - so if we're in "master/slave" world - you have "workers", and a master implementation (what is now the manager) - and it will, among other things, hold the "roles" of the workers - some workers are slaves, 1 is master.

Regardless of this point - the terminology comment I said stands;
- consumer/producer is not appropriate here and doesn't mix with master anyways
- master/slave can work (it doesn't matter if one of the slaves is master or you call them "workers" and one is master and the rest are held in some "slaves" list)
- leader/follower can work - then you can name the generic one "member"
- I guess coordinator/follower also makes sense if you like that one more 🤷 
- I guess other things as well

So there are many options, but I think "consumer" vs "master consumer" is not one of them, and mixing in a "manager" adds to the confusion in my opinion

Reply:
Talked offline
master/consumer -> leader/follower
projects manager -> project member, we'll have a member interface which currently only the leader is implementing it (everything that was under projects manager), at some point when MLRun not the master will become a valid scenario, we'll implement the follower implementation of the member interface